[Form Control Refresh] Debug assert may trigger crash during slider painting

lilyspiniolas opened the pull request:
#### 4465174f8db294718ffb70bba7528e9f07f09765
<pre>
[Form Control Refresh] Debug assert may trigger crash during slider painting
<a href="https://bugs.webkit.org/show_bug.cgi?id=295352">https://bugs.webkit.org/show_bug.cgi?id=295352</a>
<a href="https://rdar.apple.com/154888927">rdar://154888927</a>

Reviewed by Simon Fraser.

`hasVisibleSliderThumbDescendant` previously called `lastChildBox()`, which
asserts that the last child is a RenderBox. This is not always true, which
lead to the asssert triggering a crash.

Now, we attempt to downcast `lastChild()` to a RenderBox rather than calling
`lastChildBox()` to avoid hitting any asserts.

Tested by fast/forms/range/slider-mouse-events.html.

* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::hasVisibleSliderThumbDescendant):

Canonical link: <a href="https://commits.webkit.org/296947@main">https://commits.webkit.org/296947@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05fdf6b570b0709d744afafa612765c9893c236a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116049 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60275 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111990 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38273 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83650 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99083 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64092 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23579 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17229 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59845 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17274 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118840 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37067 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92621 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95350 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92445 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23566 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37429 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15168 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32949 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36961 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42432 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36623 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
You could say "tested by `fast/forms/range/slider-mouse-events.html`" or add a test specifically for this.

lilyspiniolas committed:
[Form Control Refresh] Debug assert may trigger crash during slider painting
https://bugs.webkit.org/show_bug.cgi?id=295352
rdar://154888927

Reviewed by Simon Fraser.

`hasVisibleSliderThumbDescendant` previously called `lastChildBox()`, which
asserts that the last child is a RenderBox. This is not always true, which
lead to the asssert triggering a crash.

Now, we attempt to downcast `lastChild()` to a RenderBox rather than calling
`lastChildBox()` to avoid hitting any asserts.

Tested by fast/forms/range/slider-mouse-events.html.

* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::hasVisibleSliderThumbDescendant):

Canonical link: https://commits.webkit.org/296947@main
